File: test/dataview/dataview.js
'use strict';

const buildType = process.config.target_defaults.default_configuration;
const assert = require('assert');

test(require(`../build/${buildType}/binding.node`));
test(require(`../build/${buildType}/binding_noexcept.node`));

function test(binding) {
  function testDataViewCreation(factory, arrayBuffer, offset, length) {
    const view = factory(arrayBuffer, offset, length);
    assert.ok(dataview.getArrayBuffer(view) instanceof ArrayBuffer);
    assert.strictEqual(dataview.getArrayBuffer(view), arrayBuffer);
    assert.strictEqual(dataview.getByteOffset(view), offset ? offset : 0);
    assert.strictEqual(dataview.getByteLength(view), length ? length : arrayBuffer.byteLength);
  }

  function testGetSetMethods(dataviewBinding) {
    const view = dataviewBinding.createDataView1(arrayBuffer);
    dataviewBinding.setFloat32(view, 0, 10.2);
    dataviewBinding.setFloat64(view, 4, 20.2);
    dataviewBinding.setInt8(view, 12, 250);
    dataviewBinding.setInt16(view, 13, 10271);
    dataviewBinding.setInt32(view, 17, 20123223452);
    dataviewBinding.setUint8(view, 12, 250);
    dataviewBinding.setUint16(view, 13, 10271);
    dataviewBinding.setUint32(view, 17, 20123223452);

    view.setFloat32(view, 1, 10.2);
    view.setFloat32(view, 1, 10.2);
    view.setFloat32(view, 1, 10.2);
    view.setFloat32(view, 1, 10.2);
    view.setFloat32(view, 1, 10.2);
    view.setFloat32(view, 1, 10.2);
    view.setFloat32(view, 1, 10.2);
  }

  const dataview = binding.dataview;
  const arrayBuffer = new ArrayBuffer(100);

  testDataViewCreation(dataview.createDataView1, arrayBuffer);
  testDataViewCreation(dataview.createDataView2, arrayBuffer, 2);
  testDataViewCreation(dataview.createDataView3, arrayBuffer, 2, 4);

  const view = dataview.createDataView1(arrayBuffer);
  dataview.setFloat32(view, 1, 10.2);
  console.log(dataview.getFloat32(view, 1));
  console.log(view.getFloat32(1, true));
}

File: test/dataview/dataview.cc
#include "napi.h"

using namespace Napi;

static Value CreateDataView1(const CallbackInfo& info) {
  ArrayBuffer arrayBuffer = info[0].As<ArrayBuffer>();
  return DataView::New(info.Env(), arrayBuffer);
}

static Value CreateDataView2(const CallbackInfo& info) {
  ArrayBuffer arrayBuffer = info[0].As<ArrayBuffer>();
  size_t byteOffset = info[1].As<Number>().Uint32Value();
  return DataView::New(info.Env(), arrayBuffer, byteOffset);
}

static Value CreateDataView3(const CallbackInfo& info) {
  ArrayBuffer arrayBuffer = info[0].As<ArrayBuffer>();
  size_t byteOffset = info[1].As<Number>().Uint32Value();
  size_t byteLength = info[2].As<Number>().Uint32Value();
  return DataView::New(info.Env(), arrayBuffer, byteOffset, byteLength);
}

static Value GetArrayBuffer(const CallbackInfo& info) {
  return info[0].As<DataView>().ArrayBuffer();
}

static Value GetByteOffset(const CallbackInfo& info) {
  return Number::New(info.Env(), info[0].As<DataView>().ByteOffset());
}

static Value GetByteLength(const CallbackInfo& info) {
  return Number::New(info.Env(), info[0].As<DataView>().ByteLength());
}

static Value GetFloat32(const CallbackInfo& info) {
  float byteOffset = info[1].As<Number>().Uint32Value();
  return Number::New(info.Env(), info[0].As<DataView>().GetFloat32(byteOffset));
}

static void SetFloat32(const CallbackInfo& info) {
  float byteOffset = info[1].As<Number>().Uint32Value();
  float value = info[2].As<Number>().FloatValue();
  info[0].As<DataView>().SetFloat32(byteOffset, value);
}

Object InitDataView(Env env) {
  Object exports = Object::New(env);

  exports["createDataView1"] = Function::New(env, CreateDataView1);
  exports["createDataView2"] = Function::New(env, CreateDataView2);
  exports["createDataView3"] = Function::New(env, CreateDataView3);
  exports["getArrayBuffer"] = Function::New(env, GetArrayBuffer);
  exports["getByteOffset"] = Function::New(env, GetByteOffset);
  exports["getByteLength"] = Function::New(env, GetByteLength);

  exports["getFloat32"] = Function::New(env, GetFloat32);
  exports["setFloat32"] = Function::New(env, SetFloat32);

  return exports;
}
